activitysim/core/util.py

```diff
@@ -371,6 +379,79 @@ def assign_in_place(df, df2):
 
     df[new_columns] = df2[new_columns]
 
+    for c in new_columns:
+        if pd.api.types.is_object_dtype(df[c]):
+            df[c] = df[c].astype("category")
+
+    auto_opt_pd_dtypes(df, downcast_int, downcast_float, inplace=True)
```

Comment:
By running this at the end of the assign-in-place func, it seems like we're first creating a large, inefficient dataframe with poorly optimized data types, and then trying to compress it down by using more efficient types.  In many cases this is probably fine, the data types are not really that big and it's keeping the size down on long-term stored data, but in corner cases this may make the memory usage worse instead of better

Reply:
Correct, auto_opt_pd_dtypes() compresses variables after they are created in, e.g., inputs, annotations, UECs. It's a mask and shortcut rather than changing how activitysim creates variables, i.e., eval().  For this instance, are you suggesting applying auto_opt_pd_dtypes() to df2, rather than df?  When you say "in corner cases", do you mean how pandas converts data types internally and it's possible for it to consume more memory?

Reply:
When I wrote this I was talking here about how the basic `eval` code creates object type variables, and wouldn't it be better if it didn't.  But I think you're right not to touch that yet, it's likely to get messy and be not-so backwards compatible, which we don't want to do (yet)